Teensy 3.0 / LC link-time RAM overflow (.dmabuffers / .bss) across all examples

## Summary

After consuming the `fbuild` library-selector fix (`fbuild` 2.2.3 — tracked in #2403, now closed), Teensy 3.0 and Teensy LC CI builds still fail in CI, but with a new and different signature. The previous "multiple definition of `FastLED`" link failure is gone; these are now pure RAM-overflow link failures.

This is a code-size / static-allocation regression against the tiny Teensy 3.0 (16 KB RAM) and Teensy LC (8 KB RAM) targets, not an `fbuild` bug.

## Failing CI runs (master, 8eb52e168fded843b26f18698af7b5997d685863)

- `teensy30` board build: https://github.com/FastLED/FastLED/actions/runs/24901454256
- `teensyLC` board build: https://github.com/FastLED/FastLED/actions/runs/24901454197
- `check_teensy30_size` / `check_teensylc_size` also fail (no `build_info.json` produced — downstream of the link failure)

## Observed failures

### Teensy 3.0 (16 KB RAM)
Every example tried fails with a `.dmabuffers` overflow of essentially a fixed amount (~22.4 KB — larger than the entire chip RAM):

```
ld: firmware.elf section `.dmabuffers' will not fit in region `RAM'
ld: region `RAM' overflowed by 22404 bytes   (AnalogOutput)
ld: region `RAM' overflowed by 22400 bytes   (AudioFftParity, BeatDetection, ParallelSPI, PinMode)
ld: region `RAM' overflowed by 22424 bytes   (RGBCalibrate)
ld: region `RAM' overflowed by 22412 bytes   (Test)
```

The near-constant ~22 400 byte overflow across unrelated examples strongly suggests a **single fixed-size `.dmabuffers` static allocation** that is sized for a larger Teensy (4.x) and not gated behind a per-platform size budget for Teensy 3.0.

### Teensy LC (8 KB RAM)
`.bss` overflow, smaller and varying per example:

```
ld: firmware.elf section `.bss' will not fit in region `RAM'
ld: region `RAM' overflowed by 1708–2424 bytes   (across multiple examples)
```

### Reference: passing Teensy boards
- `teensy31` (64 KB RAM): PASS
- `teensy32` (64 KB RAM): PASS
- `teensy41` board build: status differs across jobs; `check_teensy41_size` fails downstream of the global `check_*_size` breakage, but the board build itself appears healthy

This rules out a generic Teensy breakage — only the RAM-starved 3.0 / LC fail, with the 3.0 failing by a single constant amount.

## Suspected root cause

A static `.dmabuffers` (likely inside one of the Teensy-family clockless/SPI drivers) is sized for Teensy 4.x and is linked in unconditionally for Teensy 3.0 / LC as well. The fix is likely one of:

1. Gate the large `.dmabuffers` allocation behind a size-budget guard so Teensy 3.0 / LC do not instantiate it.
2. Pick a smaller DMA buffer size on Teensy 3.x / LC (they cannot support the Teensy-4-sized buffer anyway at 16 KB / 8 KB total RAM).
3. If the code path is genuinely unusable on Teensy 3.0 / LC, stub it out for those targets instead of carrying a dead 22 KB buffer.

## Minimal repro

```bash
./compile --no-interactive --no-parallel --max-failures 1 teensy30 AnalogOutput
./compile --no-interactive --no-parallel --max-failures 1 teensyLC  Blink
```

## Related

- #2403 (closed): `fbuild` library-selector fix consumed — resolves the previous `multiple definition of FastLED` symptom
- #2398: original Teensy 3.x/LC CI regression issue — can be narrowed to this RAM-overflow follow-up
- `FastLED/fbuild#199` (closed): upstream `fbuild` fix
- `FastLED/fbuild#198` (merged): upstream fix PR

## Exit criteria

- `teensy30 all` and `teensyLC all` board builds pass in CI through the `fbuild` path
- Any large `.dmabuffers` static allocation is either sized per-target or not instantiated on Teensy 3.x / LC
- Regression coverage keeps Teensy 3.0 / LC green

--- a/ci/tests/test_sketch_filter.py
+++ b/ci/tests/test_sketch_filter.py
@@ -686,6 +686,43 @@ class TestAudioInputFilter:
             )
 
 
+class TestBeatDetectionFilter:
+    """Test BeatDetection example filter behavior."""
+
+    def test_beat_detection_skips_small_teensy_boards(self) -> None:
+        """BeatDetection uses UIAudio and should skip low-memory Teensy boards."""
+        beat_detection_path = (
+            Path(__file__).parent.parent.parent
+            / "examples"
+            / "BeatDetection"
+            / "BeatDetection.ino"
+        )
+        sketch_filter = parse_filter_from_sketch(beat_detection_path)
+
+        assert sketch_filter is not None, "BeatDetection should have a @filter directive"
+
+        cases = [
+            (
+                Board(board_name="teensy30", platform="teensy", framework="arduino"),
+                True,
+            ),
+            (
+                Board(board_name="teensylc", platform="teensy", framework="arduino"),
+                True,
+            ),
+            (
+                Board(board_name="teensy41", platform="teensy", framework="arduino"),
+                False,
+            ),
+        ]
+
+        for board, should_skip_expected in cases:
+            should_skip, reason = should_skip_sketch(board, sketch_filter)
+            assert should_skip is should_skip_expected, (
+                f"unexpected BeatDetection filter result for {board.board_name}: {reason}"
+            )
+
+
 class TestMemoryTierMatching:
     """Test three-tier memory classification: low < large < huge."""
 

--- a/ci/tests/test_sketch_filter.py
+++ b/ci/tests/test_sketch_filter.py
@@ -645,6 +645,47 @@ class TestPintestFilter:
         assert "atmega8" in reason.lower() or "board" in reason.lower()
 
 
+class TestAudioInputFilter:
+    """Test AudioInput example filter behavior."""
+
+    def test_audio_input_skips_small_teensy_boards(self) -> None:
+        """AudioInput requires Teensy boards with enough RAM for PJRC Audio."""
+        audio_input_path = (
+            Path(__file__).parent.parent.parent
+            / "examples"
+            / "AudioInput"
+            / "AudioInput.ino"
+        )
+        sketch_filter = parse_filter_from_sketch(audio_input_path)
+
+        assert sketch_filter is not None, "AudioInput should have a @filter directive"
+
+        cases = [
+            (
+                Board(board_name="teensy30", platform="teensy", framework="arduino"),
+                True,
+            ),
+            (
+                Board(board_name="teensylc", platform="teensy", framework="arduino"),
+                True,
+            ),
+            (
+                Board(board_name="teensy41", platform="teensy", framework="arduino"),
+                False,
+            ),
+            (
+                Board(board_name="esp32dev", platform="esp32", framework="arduino"),
+                False,
+            ),
+        ]
+
+        for board, should_skip_expected in cases:
+            should_skip, reason = should_skip_sketch(board, sketch_filter)
+            assert should_skip is should_skip_expected, (
+                f"unexpected AudioInput filter result for {board.board_name}: {reason}"
+            )
+
+
 class TestMemoryTierMatching:
     """Test three-tier memory classification: low < large < huge."""
 

--- a/examples/Animartrix/Animartrix.ino
+++ b/examples/Animartrix/Animartrix.ino
@@ -44,6 +44,10 @@ Performence notes @64x64:
 
 #include "FastLED.h"
 
+#if defined(FL_IS_TEENSY)
+// Keep fbuild's library scanner aware of PJRC Audio sources for Teensy.
+#include <Audio.h>
+#endif
 
 // DRAW TIME: 7ms
 

--- a/examples/AnimartrixRing/AnimartrixRing.ino
+++ b/examples/AnimartrixRing/AnimartrixRing.ino
@@ -11,6 +11,11 @@
 // build system
 #include "FastLED.h"
 
+#if defined(FL_IS_TEENSY)
+// Keep fbuild's library scanner aware of PJRC Audio sources for Teensy.
+#include <Audio.h>
+#endif
+
 #include "fl/math/math.h"
 #include "fl/math/screenmap.h"
 #include "fl/ui/ui.h"

--- a/examples/Audio/Audio.ino
+++ b/examples/Audio/Audio.ino
@@ -6,6 +6,11 @@
 
 #include <FastLED.h>
 
+#if defined(FL_IS_TEENSY)
+// Keep fbuild's library scanner aware of PJRC Audio sources for Teensy.
+#include <Audio.h>
+#endif
+
 #if !SKETCH_HAS_LARGE_MEMORY
 void setup() {}
 void loop() {}

--- a/examples/AudioInput/AudioInput.ino
+++ b/examples/AudioInput/AudioInput.ino
@@ -1,4 +1,4 @@
-// @filter: (platform is teensy) or (platform is esp32)
+// @filter: ((platform is teensy) and (memory is large)) or (platform is esp32)
 
 // I2S Audio Input Example
 // This example demonstrates using I2S audio input to control FastLED strips

--- a/examples/AudioInput/TeensyAudioInput.h
+++ b/examples/AudioInput/TeensyAudioInput.h
@@ -4,6 +4,9 @@
 
 #pragma once
 
+// Keep fbuild's library scanner aware of PJRC Audio sources for Teensy.
+#include <Audio.h>
+
 #include "fl/audio/audio_input.h"
 
 // ============================================================================

--- a/examples/AudioReactive/AudioReactive.ino
+++ b/examples/AudioReactive/AudioReactive.ino
@@ -5,6 +5,11 @@
 
 #include "FastLED.h"
 
+#if defined(FL_IS_TEENSY)
+// Keep fbuild's library scanner aware of PJRC Audio sources for Teensy.
+#include <Audio.h>
+#endif
+
 #define NUM_LEDS 60
 #define LED_PIN 2
 

--- a/examples/AudioUrl/AudioUrl.ino
+++ b/examples/AudioUrl/AudioUrl.ino
@@ -11,6 +11,11 @@
 //   fastled
 
 #include <FastLED.h>
+#if defined(FL_IS_TEENSY)
+// Keep fbuild's library scanner aware of PJRC Audio sources for Teensy.
+#include <Audio.h>
+#endif
+
 #include "fl/asset/asset.h"
 #include "fl/ui/ui.h"
 #include "fl/stl/url.h"

--- a/examples/BeatDetection/BeatDetection.ino
+++ b/examples/BeatDetection/BeatDetection.ino
@@ -5,7 +5,14 @@
 /// Demonstrates beat detection using the Processor facade.
 /// Visualizes beats and tempo on an LED strip - flashes on beat detection.
 
+// @filter: (memory is large)
+
 #include <FastLED.h>
+
+#if defined(FL_IS_TEENSY)
+// Keep fbuild's library scanner aware of PJRC Audio sources for Teensy.
+#include <Audio.h>
+#endif
 
 #if !SKETCH_HAS_LARGE_MEMORY
 // Platform does not have enough memory

--- a/examples/ElPanelReactive/ElPanelReactive.ino
+++ b/examples/ElPanelReactive/ElPanelReactive.ino
@@ -15,6 +15,11 @@
 // @filter: (mem is large)
 
 #include <FastLED.h>
+#if defined(FL_IS_TEENSY)
+// Keep fbuild's library scanner aware of PJRC Audio sources for Teensy.
+#include <Audio.h>
+#endif
+
 #include "el_panel.h"
 #include "fl/math/filter/filter.h"
 #include "fl/math/math.h"

--- a/examples/Sailboat/Sailboat.ino
+++ b/examples/Sailboat/Sailboat.ino
@@ -11,6 +11,11 @@
 
 #include <FastLED.h>
 
+#if defined(FL_IS_TEENSY)
+// Keep fbuild's library scanner aware of PJRC Audio sources for Teensy.
+#include <Audio.h>
+#endif
+
 #include "fl/channels/config.h"
 #include "fl/chipsets/chipset_timing_config.h"
 #include "fl/fx/1d/perlin_particle_punch.h"

--- a/src/fl/audio/audio_input.cpp.hpp
+++ b/src/fl/audio/audio_input.cpp.hpp
@@ -15,23 +15,16 @@
 
 // First check for Teensy (before Arduino, since Teensy is Arduino-compatible)
 // IWYU pragma: begin_keep
-#include "platforms/arm/teensy/is_teensy.h" // ok platform headers
+#include "platforms/arm/teensy/audio_input_teensy_config.h" // ok platform headers
 // IWYU pragma: end_keep
-#ifndef FASTLED_USES_TEENSY_AUDIO_INPUT
-  #if defined(FL_IS_TEENSY)
-    #define FASTLED_USES_TEENSY_AUDIO_INPUT 1
-  #else
-    #define FASTLED_USES_TEENSY_AUDIO_INPUT 0
-  #endif
-#endif
 
 #ifndef FASTLED_USES_ARDUINO_AUDIO_INPUT
   #if defined(FL_IS_ESP32) && !defined(FL_IS_ESP8266)
     #define FASTLED_USES_ARDUINO_AUDIO_INPUT 0
   #elif defined(FL_IS_WASM)
     #define FASTLED_USES_ARDUINO_AUDIO_INPUT 0
-  #elif FASTLED_USES_TEENSY_AUDIO_INPUT
-    // Teensy uses its own audio implementation, not generic Arduino
+  #elif defined(FL_IS_TEENSY)
+    // Teensy uses the PJRC Audio backend when enabled, never generic Arduino I2S.
     #define FASTLED_USES_ARDUINO_AUDIO_INPUT 0
   #elif FL_HAS_INCLUDE(<Arduino.h>)
     #define FASTLED_USES_ARDUINO_AUDIO_INPUT 1
@@ -64,7 +57,10 @@
 // Include platform-specific audio input implementation
 // IWYU pragma: begin_keep
 #if FASTLED_USES_TEENSY_AUDIO_INPUT
-#include "platforms/arm/teensy/audio_input_teensy.h" // ok platform headers
+#define FASTLED_TEENSY_AUDIO_INPUT_HEADER                                    \
+    "platforms/arm/teensy/audio_input_teensy.h"
+#include FASTLED_TEENSY_AUDIO_INPUT_HEADER
+#undef FASTLED_TEENSY_AUDIO_INPUT_HEADER
 #elif FASTLED_USES_ARDUINO_AUDIO_INPUT
 #include "platforms/arduino/audio_input.hpp" // ok platform headers
 #elif FASTLED_USES_ESP32_AUDIO_INPUT

--- a/src/platforms/arm/teensy/_build.cpp.hpp
+++ b/src/platforms/arm/teensy/_build.cpp.hpp
@@ -7,7 +7,13 @@
 // Root directory implementations (alphabetical order)
 
 // begin current directory includes
-#include "platforms/arm/teensy/audio_input_teensy.cpp.hpp"
+#include "platforms/arm/teensy/audio_input_teensy_config.h"
+#if FASTLED_USES_TEENSY_AUDIO_INPUT
+#define FASTLED_TEENSY_AUDIO_INPUT_IMPL                                      \
+    "platforms/arm/teensy/audio_input_teensy.cpp.hpp"
+#include FASTLED_TEENSY_AUDIO_INPUT_IMPL
+#undef FASTLED_TEENSY_AUDIO_INPUT_IMPL
+#endif
 #include "platforms/arm/teensy/init_teensy4.cpp.hpp"
 #include "platforms/arm/teensy/semaphore_teensy.cpp.hpp"
 

--- a/src/platforms/arm/teensy/audio_input_teensy.h
+++ b/src/platforms/arm/teensy/audio_input_teensy.h
@@ -9,17 +9,16 @@
 #include "fl/stl/vector.h"
 #include "fl/stl/span.h"
 #include "fl/stl/shared_ptr.h"
-#include "platforms/arm/teensy/is_teensy.h"
+#include "platforms/arm/teensy/audio_input_teensy_config.h"
 
 // Detect Teensy Audio Library availability
-#if defined(FL_IS_TEENSY) && FL_HAS_INCLUDE(<Audio.h>)
-#define TEENSY_AUDIO_LIBRARY_AVAILABLE 1
+#if TEENSY_AUDIO_LIBRARY_AVAILABLE
 // IWYU pragma: begin_keep
-#include <Audio.h>  // ok include
+#define FASTLED_TEENSY_AUDIO_LIBRARY_HEADER <Audio.h>
+#include FASTLED_TEENSY_AUDIO_LIBRARY_HEADER
+#undef FASTLED_TEENSY_AUDIO_LIBRARY_HEADER
 #include "fl/stl/noexcept.h"
 // IWYU pragma: end_keep
-#else
-#define TEENSY_AUDIO_LIBRARY_AVAILABLE 0
 #endif
 
 namespace fl {

--- a/src/platforms/arm/teensy/audio_input_teensy_config.h
+++ b/src/platforms/arm/teensy/audio_input_teensy_config.h
@@ -1,0 +1,24 @@
+#pragma once
+
+#include "fl/stl/has_include.h"
+#include "platforms/arm/teensy/is_teensy.h"
+
+// Teensy LC and Teensy 3.0 do not have enough RAM for the PJRC Audio
+// library's static DMA buffers in normal FastLED example builds.
+#if !defined(FASTLED_USES_TEENSY_AUDIO_INPUT)
+  #if defined(FL_IS_TEENSY) && !defined(FL_IS_TEENSY_LC) &&                  \
+      !defined(FL_IS_TEENSY_30) && FL_HAS_INCLUDE(<Audio.h>)
+    #define FASTLED_USES_TEENSY_AUDIO_INPUT 1
+  #else
+    #define FASTLED_USES_TEENSY_AUDIO_INPUT 0
+  #endif
+#endif
+
+#if !defined(TEENSY_AUDIO_LIBRARY_AVAILABLE)
+  #if FASTLED_USES_TEENSY_AUDIO_INPUT && defined(FL_IS_TEENSY) &&            \
+      FL_HAS_INCLUDE(<Audio.h>)
+    #define TEENSY_AUDIO_LIBRARY_AVAILABLE 1
+  #else
+    #define TEENSY_AUDIO_LIBRARY_AVAILABLE 0
+  #endif
+#endif

--- a/src/platforms/audio.h
+++ b/src/platforms/audio.h
@@ -2,11 +2,11 @@
 #pragma once
 
 #include "fl/stl/has_include.h"
-#include "platforms/arm/teensy/is_teensy.h"
+#include "platforms/arm/teensy/audio_input_teensy_config.h"
 
 // Check for Teensy with Audio Library first
 #if !defined(FASTLED_HAS_AUDIO_INPUT)
-  #if defined(FL_IS_TEENSY) && FL_HAS_INCLUDE(<Audio.h>)
+  #if FASTLED_USES_TEENSY_AUDIO_INPUT && TEENSY_AUDIO_LIBRARY_AVAILABLE
     #define FASTLED_HAS_AUDIO_INPUT 1
   #endif
 #endif

